Hints, Problems and Flag functions

KreativeThinker opened the pull request:
Adds functionality for:
- [x] Flag checking
- [x] Hints
- [x] Problems

`get_team_id()` needs to be modified to retrieve the team ID from cookies.
` flag_get` For testing purposes only. flag to be passed in the body of `POST` function. Use commented function

KreativeThinker commented:
Added endpoint for team/list. Error in creating a user which requires further examination. Awaiting user authentication and the addition of point management to add the leaderboard and the team creation functionality.

mradigen commented on src/pwncore/routes/team.py:
```diff
@@ -9,9 +11,12 @@
-    return [{"team_name": "CID Squad"}, {"team_name": "Astra"}]
+async def team_list(response: Response):
+    teams = await Team.all().values()
+    if not teams:
```
No need to check for teams being empty, if there are no teams return the empty list.

mradigen commented on src/pwncore/routes/team.py:
```diff
@@ -20,7 +25,12 @@ async def team_login():
     return {"status": "logged in!"}
 
 
+# Unable to test as adding users returns an error
```
The model `User.save` method has a bug: It is not checking whether `members` fields exists on the Team before trying to filter for it.

In `src/pwncore/models/user.py`, change:
```py
- if self.team is not None:
+ if self.team is not None and hasattr(self.team, "members"):
```

mradigen commented on src/pwncore/routes/__init__.py:
```diff
@@ -1,10 +1,67 @@
+        await Team.create(
+            name="Triple A battery" + uuid.uuid4().hex, secret_hash="chotiwali"
+        )
+        await User.create(name="Aditya Jyoti", team_id=2)
```
Given the User model is fixed, you'll have to provide the tag (reg. no.), email, and phone number for the user to be created.

mradigen commented on src/pwncore/routes/team.py:
```diff
@@ -20,7 +25,12 @@ async def team_login():
-    # Get team members from team_id
-    return [{"name": "ABC", "user_id": 3432}, {"name": "DEF", "user_id": 3422}]
+async def team_members(team_id: int, response: Response):
+    team = await Team.get_or_none(id=team_id).values_list("members_id", flat=True)
+    if not team:
+        response.status_code = 404
+        return {"msg_code": config.msg_codes["team_not_found"]}
+    members = [await User.get(id=x) for x in team]
```
Filter directly for team_id in User instead:
```py
members = await User.filter(team_id=team_id).values()
```

WizzyGeek commented on src/pwncore/models/ctf.py:
```diff
@@ -48,7 +48,9 @@ class Meta:
     team: fields.ForeignKeyRelation[Team] = fields.ForeignKeyField("models.Team")
-    hint: fields.ForeignKeyRelation[Hint] = fields.ForeignKeyField("models.Hint")
+    hint: fields.ForeignKeyRelation[Hint] = fields.ForeignKeyField(
+        "models.Hint", related_name="hints"
```
why is related_name added here?
Why is Hint.hints pointing to ViewedHint records?
Remove if not intended

WizzyGeek commented on src/pwncore/models/user.py:
```diff
@@ -29,7 +29,7 @@ class User(Model):
         # TODO: Insert/Update in one query
         # Reason why we dont use pre_save: overhead, ugly
-        if self.team is not None:
+        if self.team is not None and hasattr(self.team, "members"):
```
Give me a minimal repro if possible btw

KreativeThinker replied:
https://github.com/lugvitc/pwncore/pull/10#discussion_r1398646209

There was an error creating new user s which was resolved by adding that statement.

KreativeThinker commented:
- [x] Changed routes.
- [x] Switched to pydantic serialization for outputs.
- [x] Switched to .exists
- [x] Removed function to view all viewed hints (required?)
- [x] Added error code `{ctf_solved: 12}`
- [ ] Remove init_db
- [ ] Rewrite `hint_get`
- [ ] Refactoring

WizzyGeek commented on src/pwncore/models/ctf.py:
```diff
@@ -38,7 +43,7 @@ class Meta:
     team: fields.ForeignKeyRelation[Team] = fields.ForeignKeyField("models.Team")
     problem: fields.ForeignKeyRelation[Problem] = fields.ForeignKeyField(
-        "models.Problem"
+        "models.Problem", related_name="problem"
```
related_name shouldn't be here

WizzyGeek commented:
What happened with the git history here?

Well anyway code looks good, nice work @KreativeThinker and @mradigen

KreativeThinker commented:
Need to make some changes for consistency. Also, @mradigen has recommended using custom json outputs. Should I switch to the recommended model or proceed with the current pydantic model?

mradigen commented:
> Need to make some changes for consistency. Also, @mradigen has recommended using custom json outputs. Should I switch to the recommended model or proceed with the current pydantic model?

Feel free to disregard my suggestion, and if it's convenient for you, proceed with the Pydantic model. I realised the benefit is that it allows the docs to know exactly what the routes will return, simplifying frontend development.

WizzyGeek commented on src/pwncore/models/container.py:
```diff
@@ -18,9 +18,7 @@ class Container(Model):
         "models.Problem", on_delete=fields.OnDelete.NO_ACTION
     )
-    team: fields.ForeignKeyRelation[Team] = fields.ForeignKeyField(
-        "models.Team", related_name="containers"
```
Readd the related_name or remove the `containers` annotation in the Team class

WizzyGeek commented on src/pwncore/models/ctf.py:
```diff
@@ -19,15 +23,19 @@ class Problem(Model):
     order = fields.SmallIntField()  # 0, 1, 2
     problem: fields.ForeignKeyRelation[Problem] = fields.ForeignKeyField(
-        "models.Problem", related_name="hints"
+        "models.Problem"
```
Readd the `related_name`

KreativeThinker replied:
We are not using it right now. Should I still readd it?

WizzyGeek replied:
I remember some code using hints to prefetch, if it was in this PR only then get rid of its annotation in Problem

KreativeThinker replied:
All prefetch queries were replaced with automatic accessing using Model.field syntax.

WizzyGeek replied:
then just remove the ReverseRelation annotation in the Problem model

WizzyGeek commented on src/pwncore/routes/ctf/__init__.py:
```diff
@@ -10,19 +24,89 @@
+        response.status_code = 404
+        return {"msg_code": config.msg_codes["ctf_not_found"]}
+
+    status = await SolvedProblem.get_or_none(team_id=get_team_id(), problem_id=ctf_id)
```
.exists

WizzyGeek commented on src/pwncore/routes/ctf/__init__.py:
```diff
@@ -10,19 +24,89 @@
+        response.status_code = 401
+        return {"msg_code": config.msg_codes["ctf_solved"]}
+
+    check_solved = await Container.get_or_none(
```
.exists

WizzyGeek commented on src/pwncore/routes/__init__.py:
```diff
@@ -1,16 +1,11 @@
 from fastapi import APIRouter
 
 from pwncore.routes import ctf, team
+from pwncore.config import config
```
Unused import?

WizzyGeek commented on src/pwncore/routes/ctf/__init__.py:
```diff
@@ -10,19 +24,89 @@
+@router.get("/completed")
+async def completed_problem_get():
+    problems = await Problem_Pydantic.from_queryset(
+        Problem.filter(solvedproblems__team_id=get_team_id())
```
`solvedproblems__team_id` does this run? 
change it to `solvedproblem__team_id`

KreativeThinker replied:
solvedproblems is generated by tortoise automatically. It runs. I will have to read as related_name field.

WizzyGeek replied:
> solvedproblems is generated by tortoise automatically. It runs. I will have to read as related_name field.

Btw, where is this documented?

KreativeThinker replied:
It is not very well documented but the autogenerated name comes up in the errors when running without a related_name and using an inaccessible key. I can change it in the next commit if needed.

WizzyGeek replied:
no need, its fine

WizzyGeek commented on src/pwncore/routes/ctf/__init__.py:
```diff
@@ -10,19 +24,89 @@
+@router.get("/{ctf_id}/viewed_hints")
+async def viewed_problem_hints_get(ctf_id: int):
+    viewed_hints = await Hint_Pydantic.from_queryset(
+        Hint.filter(problem_id=ctf_id, viewedhints__team_id=get_team_id())
```
`viewedhints__team_id` -> `viewedhint__teami_id`

WizzyGeek replied:
or you can create and explicit relation and set `related_name` to whatever you want

WizzyGeek replied:
doesn't need to changed either, this is at your discretion

WizzyGeek reviewed (approved):
LGTM

mradigen commented:
Merging this before #6 since resolving potential conflicts with #6 is easier given it has only 8 commits.
